### ISSUE TITLE
thelper false positive with generic test functions

### ISSUE THREAD

**Issue by fredbi** [golangci/golangci-lint#6351]
### Welcome

- [x] Yes, I'm using a binary release within 2 latest [releases](https://github.com/golangci/golangci-lint/releases). Only such installations are supported.
- [x] Yes, I've searched similar [issues on GitHub](https://github.com/golangci/golangci-lint/issues) and didn't find any.
- [x] Yes, I've read the `typecheck` section of the [FAQ](https://golangci-lint.run/docs/welcome/faq/#why-do-you-have-typecheck-errors).
- [x] Yes, I've tried with the standalone [linter](https://golangci-lint.run/docs/linters/) if available (e.g., gocritic, go vet, etc.).
- [x] I agree to follow this project's [Code of Conduct](https://github.com/golangci/golangci-lint?tab=coc-ov-file)

### How did you install golangci-lint?

* official github-action
* local install go install 

### Description of the problem

I am using test functions with type parameters (generics) and hit what I believe are false
positives with the thelper linter.

I've already reported this issue to the source repository for this linter: https://github.com/kulti/thelper/issues/54

Just adding it here for added visibility for other users of this linter.

Below is a sample code that triggers a thelper issue.

> Arguably this is not a helper: this is an unexported test function and I am okay to have all the detailed line numbers reported with any failure inside this function. It looks like since I've adopted generics in my test suites, the thelper check raises a lot of false positives.

Please advise.

### Version of golangci-lint

v2.8.0

thelper version
v0.7.1

### Configuration

<details>

```console
version: "2"
linters:
  default: all
  disable:
    - depguard
    - err113
    - funlen
    - godox
    - gomoddirectives
    - exhaustruct
    - nlreturn
    - nonamedreturns
    - noinlineerr
    - paralleltest
    - recvcheck
    - testpackage
    - tparallel
    - varnamelen
    - whitespace
    - wrapcheck
    - wsl
    - wsl_v5
  settings:
    dupl:
      threshold: 200
    goconst:
      min-len: 2
      min-occurrences: 3
    cyclop:
      max-complexity: 22
    gocyclo:
      min-complexity: 22
    gocognit:
      min-complexity: 35
    exhaustive:
      default-signifies-exhaustive: true
      default-case-required: true
    lll:
      line-length: 180
  exclusions:
    generated: lax
    presets:
      - comments
      - common-false-positives
      - legacy
      - std-error-handling
    paths:
      - third_party$
      - builtin$
      - examples$
formatters:
  enable:
    - gofumpt
    - goimports
  settings:
    # local prefixes regroup imports from these packages
    goimports:
      local-prefixes:
        - github.com/go-openapi
  exclusions:
    generated: lax
    paths:
      - third_party$
      - builtin$
      - examples$
issues:
  # Maximum issues count per one linter.
  # Set to 0 to disable.
  # Default: 50
  max-issues-per-linter: 0
  # Maximum count of issues with the same text.
  # Set to 0 to disable.
  # Default: 3
  max-same-issues: 0
```

</details>

### Go environment

go version go1.25.5 linux/amd64

### Verbose output of running

<details>

```console
$ golangci-lint cache clean
$ golangci-lint run -v
# paste output here
```

</details>

### A minimal reproducible example or link to a public repository

``` go
//nolint:thelper // false positive: this is not a helper
func testAllMarshalAs[Doc Text, Object any](value Object, jazon Doc, success bool) func(*testing.T) {
        return func(t *testing.T) {
                t.Run("with JSONMarshalAsT", testJSONMarshalAsT(value, jazon, success))

                t.Run("with JSONUnmarshalAsT", testJSONUnmarshalAsT(value, jazon, success))
        }
}
``` 

### Validation

- [x] Yes, I've included all information above (version, config, etc.).

### Supporter

- [ ] I am a sponsor/backer through [GitHub](https://github.com/sponsors/golangci/) or [OpenCollective](https://opencollective.com/golangci-lint)

**Comment by ldez** [comment]
Hello,


This is neither a false positive nor related to generics.


<details>
<summary>the source</summary>

```go
package main

import "testing"

type (
	Text any
)

// with generics
func oneTestAllMarshalAs[Doc Text, Object any](value Object, jazon Doc, success bool) func(*testing.T) {
	return func(t *testing.T) {
		t.Run("with JSONMarshalAsT", testJSONMarshalAsT(value, jazon, success))

		t.Run("with JSONUnmarshalAsT", testJSONUnmarshalAsT(value, jazon, success))
	}
}

// without generics
func twoTestAllMarshalAs(value any, jazon any, success bool) func(*testing.T) {
	return func(t *testing.T) {
		t.Run("with JSONMarshalAsT", testJSONMarshalAsT(value, jazon, success))

		t.Run("with JSONUnmarshalAsT", testJSONUnmarshalAsT(value, jazon, success))
	}
}

func testJSONUnmarshalAsT(value any, jazon any, success bool) func(t *testing.T) {
	return func(t *testing.T) {}
}

func testJSONMarshalAsT(value any, jazon any, success bool) func(t *testing.T) {
	return func(t *testing.T) {}
}

```

</details>


```console
$ golangci-lint run --enable-only thelper
main_test.go:13:9: test helper function should start from t.Helper() (thelper)
        return func(t *testing.T) {
               ^
main_test.go:22:9: test helper function should start from t.Helper() (thelper)
        return func(t *testing.T) {
               ^
2 issues:
* thelper: 2
```

<details>
<summary>the fix</summary>

```go
package main

import "testing"

type (
	Text any
)

// with generics
func oneTestAllMarshalAs[Doc Text, Object any](value Object, jazon Doc, success bool) func(*testing.T) {
	return func(t *testing.T) {
		t.Helper()

		t.Run("with JSONMarshalAsT", testJSONMarshalAsT(value, jazon, success))

		t.Run("with JSONUnmarshalAsT", testJSONUnmarshalAsT(value, jazon, success))
	}
}

// without generics
func twoTestAllMarshalAs(value any, jazon any, success bool) func(*testing.T) {
	return func(t *testing.T) {
		t.Helper()

		t.Run("with JSONMarshalAsT", testJSONMarshalAsT(value, jazon, success))

		t.Run("with JSONUnmarshalAsT", testJSONUnmarshalAsT(value, jazon, success))
	}
}

func testJSONUnmarshalAsT(value any, jazon any, success bool) func(t *testing.T) {
	return func(t *testing.T) {}
}

func testJSONMarshalAsT(value any, jazon any, success bool) func(t *testing.T) {
	return func(t *testing.T) {}
}

```

</details>

```console
$ golangci-lint run --enable-only thelper
0 issues.
```

A helper can be an exported function or a non-exported function.
The generics have no impact on the linter behavior.

**Comment by fredbi** [comment]
This answer is not helpful and i disagree with the conclusion but i understand that only closed issues are good issues, so let’s not bang our head more than necessary about it.

At the end of the day, this not a big problem : just one more low-value, opinionated linter to disable in my config.

**Comment by ldez** [comment]
You disagree with the fact that is not related to generics even with a proof?

You disagree with the fact, which is not an opinion, that a helper can be an exported function or a non-exported function?

But maybe by "unexported" you mean an anonymous function?

> i understand that only closed issues are good issues,

You can say that if you want, but it will not become the reality.

---

I spent time to create a concrete reproducible example based on your snippet, to analyze the topic, and wrote an answer with a proof and facts.

You are implying that I closed the issue without any kind of effort to provide something useful and you do that in a very unpleasant tone.

And you are saying negative things on a linter, which I'm not related with, and implies that golangci-lint is full of "low-value linters".

In the past, I already helped you with your issues, so I don't understand why you are so negative now.
